### DISCUSSION TITLE
Enhancement: Skip registers when assigning variables that could conflict with "used" scratch registers

### DISCUSSION THREAD

**Discussion by zatricky** [Stationeers-ic/vscode-stationeers-ic10#64] [Q&A]
**Proposal**
If icx code refers to a register directly, such as `r0`, and the sum count of "scratch registers" and regular variables does not exceed the limit of 16, reserve the used registers as "scratch registers" automatically.

**Describe the problem scenario**
I have some complex scripts in ic10 I'd like to move to icx. The aim is to make it easier to read while also making it so I can minify the end-result without losing comments and organisational whitespace. I and others have conventions to use the lower registers such as r0/r1 as scratch registers. This means these registers are normally only used for temporary values that we immediately discard.

Unfortunately, using these registers as scratch registers conflicts with the registers icx uses when converting to ic10. I'm sure there's a different way to handle this besides my suggestion - but this is the simplest thing I could think of at this time.

**Example simple non-ICX script**
This ic10 script uses two strategically-placed daylight sensors to send a signal of whether or not it is nighttime to a Logic Memory unit. This is so indoor and outdoor lights can be turned on during dawn/dusk without any period of darkness.

    alias nighttime_memory d0
    alias east_daylight_sensor d1
    alias west_daylight_sensor d2
    
    alias nighttime r4                  # regular variable
    
    l nighttime db Setting
    
    start:
    s db Setting nighttime
    bdns d0 skip_to_yield
    s nighttime_memory Setting nighttime
    skip_to_yield:
    yield
    
    l r0 east_daylight_sensor Activate  # Scratch register
    seqz nighttime r0
    l r0 west_daylight_sensor Activate
    seqz r1 r0
    and nighttime nighttime r1
    
    j start

Take note of line 5 `alias nighttime r4` and line 16 `l r0 east_daylight_sensor Activate`.

Converted into icx:

    use aliases
    alias nighttime_memory d0
    alias east_daylight_sensor d1
    alias west_daylight_sensor d2
    
    var nighttime = db.Setting          # regular variable
    
    start:
    db.Setting = nighttime
    bdns d0 skip_to_yield
    nighttime_memory.Setting = nighttime
    skip_to_yield:
    yield
    
    r0 = east_daylight_sensor.Activate #Problem
    seqz nighttime r0
    r0 = west_daylight_sensor.Activate
    seqz r1 r0
    and nighttime nighttime r1
    
    j start

Converted back into ic10:

    alias nighttime r0                  # variable using r0
    alias nighttime_memory d0
    alias east_daylight_sensor d1
    alias west_daylight_sensor d2
    l nighttime db Setting
    start:
    s db Setting nighttime
    bdns d0 skip_to_yield
    s nighttime_memory Setting nighttime
    skip_to_yield:
    yield
    l r0 east_daylight_sensor Activate  # scratch variable also using r0
    seqz nighttime r0
    l r0 west_daylight_sensor Activate
    seqz r1 r0
    and nighttime nighttime r1
    j start

When we convert the script to icx, we can no longer specify that we want the variables to use registers starting with `r4`. The `nighttime` variable ends up being assigned as `r0`, which then conflicts with the line `l r0 east_daylight_sensor Activate`. A "workaround" would be to start making extra unnecessary variables such as `daytime` or `scratch`, or to start using r15 as the scratch variable. Unfortunately without complex logics or math, such as `nighttime = not (east_daylight_sensor.Activate or west_daylight_sensor.Activate)`, these scratch registers are still very convenient.

**Desired Result**
By avoiding the conflict, in the example above, the ic10 output in line 1 would say `alias nighttime r2`, instead of `alias nighttime r0`, as both `r0` and `r1` are explicitly referenced elsewhere in the icx script.

**Comment by Traineratwot** [comment]
When you use icX you must use IcX :) 

Do not use registers in icX. this interferes with the "memory manager"
```
use aliases
alias nighttime_memory d0
alias east_daylight_sensor d1
alias west_daylight_sensor d2

var nighttime = db.Setting          # regular variable

start:
db.Setting = nighttime
bdns d0 skip_to_yield
nighttime_memory.Setting = nighttime
skip_to_yield:
yield
var b #FIX Problem
var a = east_daylight_sensor.Activate #FIX Problem
seqz nighttime a
a = west_daylight_sensor.Activate
seqz b a
and nighttime nighttime b

j start
```

Maybe I should add a warning message when using registers in icx

P.S
I don't know English well, and maybe I could misunderstand you

**Comment by zatricky** [comment]
You've understood it fine :-)

I was thinking also about minifying the ic10 result. If that's not necessarily a project goal for you, your idea of adding a warning will still be helpful to prevent this issue.
